### PR TITLE
Protected Call all Level Loads

### PR DESCRIPTION
**Pull request by lewri**
Brought on by the discovery of #2512 

**Describe what the proposed change does**
- Helps prevent a program crash where an error occurs during load.

This patch currently loses the `debug.stacktrace` call though. If we need that we'll need to use `xpcall` and some tricks to work around Lua 5.1. As a result, this currently only reports the actual line that failed.

NB: This affects all calls of  `App:loadLevel` but not the more direct `App:load` for existing saves

### CODE REVIEW

**Comment by lewri** [comment]
We do a lot of localised `pcalls` I've missed here actually. I'll WIP this

**Comment by lewri** [comment]
The crashes from preventing crashes are fixed now 😅 .

Can discuss if we want xpcall now or not for a `debug.stacktrace` or not -- some of the old code already pcalled so the error was lost anyway.

File: CorsixTH/Lua/dialogs/fullscreen/fax.lua [hunk 1]
```diff
@@ -207,16 +207,17 @@ function UIFax:choice(choice_number)
     self.ui.hospital.player_salary = self.ui.hospital.salary_offer
     if tonumber(self.ui.app.world.map.level_number) then
       local next_level = self.ui.app.world.map.level_number + 1
-      self.ui.app:loadLevel(next_level, self.ui.app.map.difficulty)
-      self.ui.app.moviePlayer:playAdvanceMovie(next_level)
+      if self.ui.app:loadLevel(_S.errors.load_level_prefix, next_level, self.ui.app.map.difficulty) then
```
Comment (lewri) [inline CorsixTH/Lua/dialogs/fullscreen/fax.lua:210]: An open fax on save doesn't crash here 👍

File: CorsixTH/Lua/app.lua [hunk 2]
```diff
@@ -671,10 +672,28 @@ function App:getAbsolutePathToLevelFile(level)
   return nil, "Level not found: " .. level
 end
 
--- Loads the specified level. If a string is passed it looks for the file with the same name
+--! Invokes a protected call of App:_loadLevel(...). See that function for more information.
+--! This function should always be called to catch errors and properly pass the
+--! error to the player
+--!param error_prefix (string) Prefixes the error relevant to what was loaded
+--! return (boolean) The outcome of the pcall
+function App:loadLevel(error_prefix, level, difficulty, level_name, level_file, level_intro, map_editor)
```
Comment (TheCycoONE) [inline CorsixTH/Lua/app.lua:680]: error_prefix should not be the first argument as it's an optional and not central to the function purpose.

Comment (lewri) [reply CorsixTH/Lua/app.lua:680]: I agree, I was being lazy.

Comment (lewri) [reply CorsixTH/Lua/app.lua]: Moved to the end of the function params, but this may have been just as bad.

File: CorsixTH/Lua/app.lua [hunk 3]
```diff
@@ -1703,18 +1722,13 @@ function App:restart()
       self.ui:addWindow(UIInformation(self.ui, { _S.information.cannot_restart }))
       return
     end
-    local status, err = pcall(self.loadLevel, self, level, difficulty, name, file, intro)
-    if not status then
-      err = "Error while loading level: " .. err
-      print(err)
-      self.ui:addWindow(UIInformation(self.ui, { err }))
-    end
+    self:loadLevel(_S.errors.load_level_prefix, level, difficulty, name, file, intro)
```
Comment (lewri) [inline CorsixTH/Lua/app.lua:1725]: I missed one. I'll just push a quick pr to fix it in a bit

**Pull request merged**